File: frontend/src/types/stayZone.ts
export type StayZoneStatus = 'proposed' | 'confirmed' | 'rejected';

export type StayZoneStrategy = 'compromise' | 'split' | 'main_cluster';

/** HOT-ZONE-TAG：机酒状态（卡上次要） */
export type StayZoneFitTag =
  | 'current_anchor'
  | 'preference_fit'
  | 'compromise'
  | 'needs_city_change';

export const STAY_ZONE_FIT_TAG_LABELS: Record<StayZoneFitTag, string> = {
  current_anchor: '当前锚点',
  preference_fit: '贴合偏好',
  compromise: '折中',
  needs_city_change: '需换城',
};

/** HOT-THEME-TAG：提示词主题（卡顶主展示；运行时，不写回 preference_tags） */
export interface PromptThemeTag {
  id: string;
  label: string;
  polarity?: 'positive' | 'negative';
  confidence?: number;
  evidence?: string;
  entity?: string;
}

export interface ThemeRef {
  id: string;
  label: string;
}

const THEME_CHIP_MAX = 4;

/** 卡顶主题：最多 4 个，溢出返回 +N */
export function stayZoneThemeChips(
  matched?: ThemeRef[] | null,
): { shown: ThemeRef[]; overflow: number } {
  const list = matched ?? [];
  if (list.length <= THEME_CHIP_MAX) return { shown: list, overflow: 0 };
  return {
    shown: list.slice(0, THEME_CHIP_MAX),
    overflow: list.length - THEME_CHIP_MAX,
  };
}

export function stayZoneFitTagLabel(tag?: StayZoneFitTag | null): string {
  if (!tag) return STAY_ZONE_FIT_TAG_LABELS.current_anchor;
  return STAY_ZONE_FIT_TAG_LABELS[tag] ?? tag;
}

export function isStayZoneBookable(zone: {
  bookable?: boolean;
  fit_tag?: StayZoneFitTag;
}): boolean {
  if (zone.bookable === false) return false;
  if (zone.fit_tag === 'needs_city_change') return false;
  return true;
}

export interface StayZoneCircleGeometry {
  type: 'circle';
  center: { lat: number; lng: number };
  radius_m: number;
}

export interface StayZonePolygonGeometry {
  type: 'polygon';
  coordinates: [number, number][];
}

export type StayZoneGeometry = StayZoneCircleGeometry | StayZonePolygonGeometry;

export interface StayZonePreferences {
  transit: number;
  minimize_hotel_moves?: boolean;
  quiet?: number;
  family_friendly?: number;
  budget?: number;
  walk_tolerance?: 'low' | 'medium' | 'high';
  safety_sensitive?: boolean;
  scenery?: number;
}

export interface RecommendedStayZone {
  id: string;
  sequence: number;
  city: string;
  label: string;
  check_in: string;
  check_out: string;
  rationale: string;
  transit_note?: string;
  strategy?: StayZoneStrategy;
  anchor_hints?: string[];
  covers_day_indices?: number[];
  status: StayZoneStatus;
  geometry?: StayZoneGeometry;
  purchase_url?: string;
  fit_tag?: StayZoneFitTag;
  bookable?: boolean;
  tag_note?: string;
  matched_themes?: ThemeRef[];
  uncovered_themes?: ThemeRef[];
}

export interface StayZoneRecommendRequest {
  trip_request: import('./tripRequest').TripRequest;
  flights: unknown[];
  itinerary?: import('./itinerary').Itinerary | null;
  preferences?: StayZonePreferences;
}

export interface StayZoneRecommendResponse {
  zones: RecommendedStayZone[];
  fetched_at: string;
  source: 'llm' | 'heuristic';
  warnings: string[];
  prompt_themes?: PromptThemeTag[];
}

export type HotelBookingStatus = 'zone_only' | 'selected' | 'booked_external';

export function defaultStayZonePreferences(): StayZonePreferences {
  return { transit: 1.0, minimize_hotel_moves: true, walk_tolerance: 'medium' };
}

export type StayZonePanelPhase = 'idle' | 'pending' | 'done';

export function getStayZonePanelPhase(intel: {
  recommended_stay_zones?: RecommendedStayZone[];
  hotels: { lat?: number; lng?: number; booking_status?: string; name?: string }[];
}): StayZonePanelPhase {
  const zones = intel.recommended_stay_zones ?? [];
  const hotelLocked = intel.hotels.some(
    (h) =>
      Boolean(h.name?.trim()) &&
      h.lat != null &&
      h.lng != null &&
      (Math.abs(h.lat) > 1e-6 || Math.abs(h.lng) > 1e-6) &&
      h.booking_status !== 'zone_only',
  );
  // P88: badge「已完成」= 已锁定具体酒店；仅片区仍为待确认
  if (hotelLocked) return 'done';
  if (zones.some((z) => z.status === 'confirmed' || z.status === 'proposed')) return 'pending';
  return 'idle';
}

export function stayZonePanelBadgeLabel(phase: StayZonePanelPhase): string {
  switch (phase) {
    case 'done':
      return '已锁酒店';
    case 'pending':
      return '待锁酒店';
    default:
      return '阶段 B';
  }
}

export function formatStayZoneDays(zone: RecommendedStayZone): string {
  const idx = zone.covers_day_indices ?? [];
  if (idx.length === 0) return `${zone.check_in} → ${zone.check_out}`;
  const days = idx.map((i) => `Day ${i + 1}`).join('、');
  return `${days} · ${zone.check_in} → ${zone.check_out}`;
}

File: backend/app/schemas/stay_zone.py
from typing import Literal, Optional

from pydantic import BaseModel, Field

from app.schemas.chat import TripRequestIn


class StayZonePreferencesIn(BaseModel):
    transit: float = 1.0
    minimize_hotel_moves: bool = True
    quiet: Optional[float] = None
    family_friendly: Optional[float] = None
    budget: Optional[float] = None
    walk_tolerance: Optional[Literal["low", "medium", "high"]] = "medium"
    safety_sensitive: bool = False
    scenery: Optional[float] = None


class StayZoneCircleGeometry(BaseModel):
    type: Literal["circle"] = "circle"
    center: dict[str, float]
    radius_m: float = 800.0


class StayZonePolygonGeometry(BaseModel):
    type: Literal["polygon"] = "polygon"
    coordinates: list[list[float]]


class ThemeTagOut(BaseModel):
    id: str
    label: str
    polarity: Literal["positive", "negative"] = "positive"
    confidence: Optional[float] = None
    evidence: Optional[str] = None
    entity: Optional[str] = None


class ThemeRefOut(BaseModel):
    id: str
    label: str


class RecommendedStayZoneOut(BaseModel):
    id: str
    sequence: int
    city: str
    label: str
    check_in: str
    check_out: str
    rationale: str
    transit_note: Optional[str] = None
    strategy: Optional[Literal["compromise", "split", "main_cluster"]] = "compromise"
    anchor_hints: list[str] = Field(default_factory=list)
    covers_day_indices: list[int] = Field(default_factory=list)
    status: Literal["proposed", "confirmed", "rejected"] = "proposed"
    geometry: Optional[StayZoneCircleGeometry | StayZonePolygonGeometry] = None
    purchase_url: Optional[str] = None
    # HOT-ZONE-TAG：机酒状态（次要）
    fit_tag: Literal[
        "current_anchor", "preference_fit", "compromise", "needs_city_change"
    ] = "current_anchor"
    bookable: bool = True
    tag_note: Optional[str] = None
    # HOT-THEME-TAG：提示词主题命中 / 未覆盖（主展示）
    matched_themes: list[ThemeRefOut] = Field(default_factory=list)
    uncovered_themes: list[ThemeRefOut] = Field(default_factory=list)


class StayZoneRecommendRequest(BaseModel):
    trip_request: TripRequestIn
    flights: list[dict] = Field(default_factory=list)
    itinerary: Optional[dict] = None
    preferences: Optional[StayZonePreferencesIn] = None


class StayZoneRecommendResponse(BaseModel):
    zones: list[RecommendedStayZoneOut]
    fetched_at: str
    source: Literal["llm", "heuristic"] = "heuristic"
    warnings: list[str] = Field(default_factory=list)
    # 运行时主题（不写回 preference_tags）
    prompt_themes: list[ThemeTagOut] = Field(default_factory=list)


class StayZoneLodgingRequest(BaseModel):
    zone_id: str
    city: str
    label: str = ""
    lat: float
    lng: float
    radius_m: float = Field(default=1200, ge=200, le=5000)
    limit: int = Field(default=8, ge=1, le=20)
    # HOT-TRIP: Trip 深链日期 / 人数 / 预算
    check_in: str = ""
    check_out: str = ""
    adults: int = Field(default=2, ge=1, le=9)
    max_price: Optional[float] = Field(default=None, ge=0)


class StayZoneLodgingCandidate(BaseModel):
    name: str
    lat: float
    lng: float
    address: Optional[str] = None
    place_id: Optional[str] = None
    rating: Optional[float] = None
    distance_m: int = 0
    coord_source: str = "serpapi_lodging"
    # HOT-RG-02: RollingGo 参考价 / 渠道链（非 Trip）
    ref_price: Optional[float] = None
    currency: Optional[str] = None
    booking_url: Optional[str] = None


class StayZoneLodgingResponse(BaseModel):
    zone_id: str
    candidates: list[StayZoneLodgingCandidate] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    status: Literal[
        "ok", "empty", "rate_limited", "provider_error", "unconfigured", "trip_first"
    ] = "ok"
    query: str = ""
    # HOT-TRIP-01: always prefer deep link when structured search unavailable
    trip_url: str = ""
    mode: Literal["rollinggo", "serp", "trip_first"] = "trip_first"
